File: 452/minimum-number-of-arrows-to-burst-balloons.py
# 氣球超寬的，從[start,end]中間射過的箭，都能把這個氣球弄破
# 要多少箭，才能把「全部的氣球」都弄破。
# 策略：以「氣球end的位置」排序，再依序處理
# 挑個氣球，射它的最end的位置，後面的氣球，只要start還在左邊，就還在「一箭」的範圍
# 如果 start 跑到前一個的end的右邊，那就無法「一箭搞定」只好再加「新箭」
class Solution:
    def findMinArrowShots(self, points: List[List[int]]) -> int:
        points.sort(key=lambda x:x[1]) # 以 右邊界 x_end 排序
        ans = 1 # 因最少有1個氣球，所以最少有1箭，放在end的位置（發射）
        end = points[0][1] # 前一箭射的位置
        for x_start, x_end in points:
            if end < x_start: # 範圍外, 需要新的一箭
                ans += 1 # 加一箭
                end = x_end # 這個難搞的氣球，右邊界當成我們要射箭的地方
        return ans

File: 452/minimum-number-of-arrows-to-burst-balloons.cpp
// LeetCode 452. Minimum Number of Arrows to Burst Balloons
// 氣球黏在牆上排一排，每個氣球超寬，射箭時，只要在 [start,end]碰到，就會破
// 最多有10萬個氣球，要將全部氣球射破，最少要射幾枝箭？
class Solution {
public:
    int findMinArrowShots(vector<vector<int>>& points) {
        sort(points.begin(), points.end()); // 以 begin 為主，小到大排好
        int N = points.size();
        int ans = 1, prevShot = points[N-1][0]; // 第1箭，射在最後1個左界
        for(int i=N-2; i>=0; i--) { // 接下來，倒著巡其他筆
            if(points[i][1]<prevShot){ // 如果「前一箭」沒能射到
                prevShot = points[i][0]; // 就射此氣球的左界
                ans++; // 多了一箭
            }
        }
        return ans;
    }
};
